int32 pos1
int32 pos2
bool begin